### DISCUSSION TITLE
How can I apply the SVG `<template>` polyfill for elements returned by HTMX?

### DISCUSSION THREAD

**Discussion by ocharles** [alpinejs/alpine#3690] [1. Help] [answered]
I have the following bit of HTML:

```html
<form hx-post="/foo">
  <button>Go</button>
</form>
```

such that `POST /foo` responds with:

```
<div x-data="{foo: true}">
  <svg>
    <template x-if="foo">
      <g></g>
    </template>
  </svg>
</div>
```

However, this makes Alpine very unhappy:

```
Alpine Expression Error: e.content is undefined

Expression: "foo"

<template x-if="foo">
```

I believe this is because the [SVG polyfill](https://github.com/alpinejs/alpine/issues/637#issuecomment-654856538) hasn't been applied. If I change `POST /foo` to return this:

```html
<div x-data="{foo: true}">
    <template x-if="foo">
      <div>OK</div>
    </template>
</div>
```

then all is well - so it does seem to be specific to `<svg>`.

Does anyone have any ideas on how I can get this working? Things I've tried:

* I tried setting `hx-on::before-swap="Alpine.stopObservingMutations()"` and `hx-on::after-swap="polyfillTemplates(event.target); Alpine.init(event.target); Alpine.startObservingMutations()"`, which now gives

```
Alpine Expression Error: e is null

Expression: "foo"

<template x-if="foo">
```

which is a bit different? But still not working :sweat_smile: 

I'm not quite sure what else to try. I may try and get HTMX to give us an event that would let apply this polyfill, but currently that doesn't exist.

**Comment by ekwoka** [comment]
you just have to run the polyfill function on the `afterswap` event from htmx

**Reply by ocharles** [reply]
Can you provide an example? As I mentioned in my post I've tried that and it doesn't work (but does produce a different error). I believe this is because the event guess too late, and Alpines mutation observer has already fired (but this is a guess)

**Reply by ekwoka** [reply]
What did you try? What error for you get?

**Reply by ocharles** [reply]
It's in my original post, towards the bottom - if that's not enough information I can try and provide a codepen example or something. Specifically see:

> I tried setting `hx-on::before-swap="Alpine.stopObservingMutations()" and hx-on::after-swap="polyfillTemplates(event.target); Alpine.init(event.target); Alpine.startObservingMutations()"`, which now gives...

**Reply by ocharles** [reply]
I should also add that:

```javascript
window.polyfillTemplates = function(parent) {
  var templates = document.querySelectorAll('svg template');
  var el, template, attribs, attrib, count, child, content;
  for (var i=0; i<templates.length; i++) {
    el = templates[i];
    template = el.ownerDocument.createElement('template');
    el.parentNode.insertBefore(template, el);
    attribs = el.attributes;
    count = attribs.length;
    while (count-- > 0) {
      attrib = attribs[count];
      template.setAttribute(attrib.name, attrib.value);
      el.removeAttribute(attrib.name);
    }
    el.parentNode.removeChild(el);
    content = template.content;
    while ((child = el.firstChild)) {
      content.appendChild(child);
    }
  }
}
```

which I think I got right.

**Comment by ekwoka** [comment]
hmm, it might be simple enough to incorporate this into the Alpine template handling logic.

I'm not sure if that would fly though.

Also, I believe the command you wanted was `Alpine.initTree`.

The error seems to be that it's not properly following the tree up to find the data context 🤔 . I'm not sure how likely that is to be a part of Alpine vs your own code...

I'd start with rethinking if the template needs to be in the SVG...

**Reply by ocharles** [reply]
For a bit more context, I'm building what is essentially a vector editor. I have a canvas where users can draw lines (they are "marking up" an image with semantic information). This actually all works fine on page load, but I was using HTMX to replace basically the entire `<body>` if the user specified they are editing a different image. A full page refresh works fine, so I do have a work around. WRT having `<template>` in the SVG, I want it for things like:

```html
<svg>
  <g>
    <template x-for="done in markUpLines">
      <line :cx .. />
    </template>
  </g>
</svg>
```

which is might convenient - giving that up would be a real shame, and I think I'd have to start managing `<line>` nodes myself in JavaScript. Doable, but this is a lot more obvious IMO.

> Also, I believe the command you wanted was `Alpine.initTree`.

Yes, that sounds likely! I'll try that and see if it fixes anything.

>  I'm not sure how likely that is to be a part of Alpine vs your own code...

Let me get back to you with a MRE and we can take it from there?

**Comment by ocharles** [comment, marked as answer]
Ok, I got it working! I believe @ekwoka spotting that I was using `init` instead of `initTree` was the problem. The following pen works: https://codepen.io/ocharles/pen/jOQdeqB?editors=1111. I need to try this on my larger code, but I'm more confident it will work now.

**Reply by ocharles** [reply]
I can confirm this works in my larger application. Having to `stop`/`start` Alpine is a shame, but probably the best I can do for now. I think if I can get HTMX to give me a pre-swap hook, that's probably the most elegant. Alternatively, Alpine could give me such a hook.
